cep-cache-output.md

```diff
@@ -0,0 +1,103 @@
+# CEP for the cache output in v1 recipes / rattler-build
+
+<table>
+<tr><td> Title </td><td> The cache output in v1 recipes / rattler-build </td>
+<tr><td> Status </td><td> In Discussion </td></tr>
+<tr><td> Author(s) </td><td> Wolf Vollprecht &ltw.vollprecht@gmail.com&gt; </td></tr>
+<tr><td> Created </td><td> Nov 27, 2024</td></tr>
+<tr><td> Updated </td><td> </td></tr>
+<tr><td> Discussion </td><td>  </td></tr>
+<tr><td> Implementation </td><td> rattler-build </td></tr>
+</table>
+
+## Abstract
+
+This CEP aims to define the cache output for v1 multi-output recipes.
+
+## Background
+
+Sometimes it is very useful to build some code once, and then split it into multiple build artifacts (such as shared library, header files, etc.). For this reason, `conda-build` has a special, implicit top-level build.
+
+There are many downsides to the behavior of `conda-build`: it's very implicit, hard to understand and hard to debug (for example, if an output is defined with the same name as the top-level recipe, this output will get the same requirements attached as the top-level).
+
+For the v1 spec we are attempting to formalize the workings of the "top-level" build. For this, we introduce a new top-level key `cache`, that has the same values as a regular output.
+
+## Specification
+
+The top-level `cache` key looks as follows:
+
+```yaml
+cache:
+  requirements:
+    build:
+      - ${{ compiler('c') }}
+      - cmake
+      - ninja
+    host:
+      - libzlib
+    # run: ...
+    # run_constraints: ...
+    # ignore_run_exports: ...
+  build:
+    # note: in the current implementation, all other keys for build
+    #       are also valid, but unused
+    script: build.sh
```

Comment:
One thing that's been a permanent annoyance with the way things have grown in conda-build is that a "global" `build.skip: ...` does not mean the entire build will get cancelled, because - in theory - the various outputs could do their own thing that doesn't rely on the package-level build. So in order to actually skip a given combination, it needs to be skipped _on all outputs_ (which quickly gets annoying in large multi-output recipes, especially for quick debugging).

If we go with the "all outputs implicitly depend on `cache:`" approach, then I think it would be good to specify that a `skip: true` on the `cache:` level completely skips the build, without having to repeat it per output.

Reply:
Hmm, I don't think the cache should have anything to do with skipping. 

The top-level can skip fine (which is intersected with the outputs).

Ie. would skip everything on win:

```
build:
  skip: [win]

outputs:
  - ... 
```